Issue 221 build results

ssciolla opened the pull request.

lsloan commented on ccm_web/server/src/api/api.admin.handler.ts:
```diff
@@ -154,39 +153,26 @@ export class AdminApiHandler {
-  async createExternalUsers (users: SectionExternalUserDto[], accountID: number): Promise<Array<CanvasUserLoginEmail>> {
+  async createExternalUsers (
+    users: SectionExternalUserDto[], accountID: number
+  ): Promise<Array<{ result: CanvasUserLoginEmail | APIErrorData, email: string }>> {
```
This is interesting!  I didn't know about this format for specifying an array of objects.

ssciolla replied:
It's basically just an inline interface. This particular interface doesn't feel particularly re-usable to me, so I just put it here for now.

lsloan commented on ccm_web/server/src/api/api.interfaces.ts:
```diff
@@ -14,8 +17,10 @@ export interface Globals {
+
 export interface APIErrorPayload {
-  canvasStatusCode: number
+  service: 'Canvas' | 'Cirrus'
```
How do you feel about using an enum here instead of service name strings?

ssciolla replied:
We can do that. I'm still contemplating whether using `APIErrorData` for both Canvas and Cirrus makes sense.

ssciolla replied:
I decided to walk back the changes to `APIErrorPayload` for the time being and instead add a separate interface and type guard for Cirrus errors. I ultimately didn't want to import stuff from `api` in `invitation`, and the nested structure of `APIErrorData`/`APIErrorPayload` felt clunky and didn't make sense for this use case. We can discuss more.

lsloan commented on ccm_web/server/src/api/api.interfaces.ts:
```diff
@@ -28,3 +33,16 @@ export interface APIErrorData {
+
+export interface ExternalEnrollmentUserData {
+  [email: string]: {
+    userCreated?: CanvasUserLoginEmail | APIErrorData | false
+    inviteResult?: CirrusInvitationResponse | APIErrorData
+    enrollment?: CanvasEnrollment | APIErrorData
```
I've not re-reviewed all the code since your walkthrough of it with me, but it seems to me that of these three properties, only `inviteResult` should ever be optional.  That's when the user creation doesn't happen, due to either error or the user exists.

ssciolla replied:
Yeah, this is something I want to look at, I just marked them all optional to get it working. We can definitely make `userCreated` required. Basically because we're accumulating, we either have to leave some things optional as we build it, or we do some other object construction where we hold parts until everything is ready. With TypeScript, it's tricky to change objects too much unless you type it at the beginning, but then it has to be fairly lenient (with optionals). I have to play around with it more.

ssciolla replied:
`userCreated` is now required.

lsloan commented on ccm_web/server/src/api/api.service.ts:
```diff
@@ -137,44 +132,62 @@ export class APIService {
+        userCreated = result
+        newUsers.push(result)
+      }
+      resultData[email] = {}
+      resultData[email].userCreated = userCreated
```
I'm not certain of the syntax, but couldn't this be… 
```suggestion
      resultData[email] = {userCreated: userCreated}
```

ssciolla replied:
Yeah, that should work, or even `resultData[email] = { userCreated }`

ssciolla replied:
Done.

lsloan commented on ccm_web/server/src/api/api.service.ts:
```diff
@@ -137,44 +132,62 @@ export class APIService {
-      inviteResults = { errors: [String(e.message)] }
+    if (newUsers.length > 0) {
+      const inviteResult = await this.invitationService.sendInvitations(newUsers.map(u => u.email))
+      Object.keys(resultData).forEach(email => {
+        resultData[email].inviteResult = inviteResult
+      })
```
I like this syntax.  I think it probably would've helped me in the code I was working on in parallel.

lsloan commented on ccm_web/server/src/api/api.utils.ts:
```diff
@@ -30,10 +31,11 @@ export function handleAPIError (error: unknown, input?: string): APIErrorPayload
     logger.error(`An error occurred while making a request to Canvas: ${JSON.stringify(error)}`)
-    return { canvasStatusCode: 500, message: 'A non-HTTP error occurred while communicating with Canvas.', failedInput: failedInput }
+    const defaultMessage = 'A non-HTTP error occurred while communicating with Canvas.'
+    return { statusCode: 500, message: defaultMessage, failedInput: failedInput, service: 'Canvas' }
```
```suggestion
    return { statusCode: HttpStatus.INTERNAL_SERVER_ERROR, message: defaultMessage, failedInput: failedInput, service: 'Canvas' }
```

lsloan reviewed (commented):
Thanks for helping with this.  I didn't expect such an extensive solution from you.  I thought you were going to address the specific thing I'd asked about.  But, I get it:  There were lots of little things that you'd have liked me to do better. Rather than have a longer, drawn-out conversation in the PR about it with me, it's more expedient if you just make the changes yourself, because you're more comfortable with TypeScript than I am.

I have just a couple of comments with suggestions for very minor things I'd do differently. Other than that I think it looks great.  I'll check out a copy and give it a run.

One more question:  Did you redo your commits?  Originally, I saw only two commits at around 16:30.  Now I see you added a commit to fix some of the errors we saw when you ran it, but all three commits were done at 17:58 and the existing ones have new hash numbers.

ssciolla commented:
> Thanks for helping with this. I didn't expect such an extensive solution from you. I thought you were going to address the specific thing I'd asked about. But, I get it: There were lots of little things that you'd have liked me to do better. Rather than have a longer, drawn-out conversation in the PR about it with me, it's more expedient if you just make the changes yourself, because you're more comfortable with TypeScript than I am.

I'm sorry for stepping on your toes. I don't think this is perfect, and I wasn't quite sure of the best solution without working it out myself a bit. I wanted to address your question while also moving the PR along, since I'll need to work with this also in the frontend to finish the issue. Hopefully the discussion and changes we make here can get us to a good solution.

> One more question: Did you redo your commits? Originally, I saw only two commits at around 16:30. Now I see you added a commit to fix some of the errors we saw when you ran it, but all three commits were done at 17:58 and the existing ones have new hash numbers.

I added one commit to resolve some of the errors we were seeing. And then I rebased to incorporate your latest commits, which I did not have. That's why you see all the commits from the same time.

lsloan commented:
I ended up bungling the handling of this PR.  I thought that I was using my git tool correctly to merge it, but instead it seems to have pulled in @ssciolla's changes without reference to his commits.  So, when I committed them to my branch, they do not reference his commits or show that he is an author.

I've apologized to Sam for this mistake.
